### ISSUE TITLE
HistoryCleanupTest failing

### ISSUE THREAD

**Issue by kthoms** [operaton/operaton#60]
**Describe the bug**
On multiple branch builds the test `HistoryCleanupTest` is currently failing in `testLessThanThresholdOutsideBatchWindowAfterMidnight()`

**To Reproduce**
Run normal Maven build. The issue might be time related. I'm running at 11 PM, and the test is doing something around midnight.
Example builds:
- https://github.com/operaton/operaton/actions/runs/11543916029/job/32128853176
- https://github.com/operaton/operaton/actions/runs/11544115859/job/32129278034

**Expected behavior**
Test should work independent on the machine's time.

**Screenshots**
![Screenshot 291](https://github.com/user-attachments/assets/fc562558-258b-4054-82ef-32501a6bd41b)

**Comment by ArneDeutsch** [comment]
I had a short look at the test class. It is operating through static methods to manipulate the time returned by the timing methods. This is inherently NOT thread safe. Assuming the tests would run concurrently the timing of the tests would be important. Not sure yet this is the reason of the fail, but it would explain them.

I refer to these calls:

```
ClockUtil.setCurrentTime(DateUtils.setMinutes(DateUtils.setHours(date, 22), 10));  //22:10
[...]
ClockUtil.setCurrentTime(DateUtils.setMinutes(DateUtils.setHours(date, 23), 10));  //23:10
```

If this assumption is true the possible solutions would be:
1. Make sure the tests are run sequentially (e.g. in the maven configuration)
2. Synchronize the tests themself to make sure they do not interfere).

**Comment by ArneDeutsch** [comment]
Checked the maven build.

`<surefire.forkCount>0.5C</surefire.forkCount>`

This means we use half the number of CPU cores as number of threads to run the tests concurrently -> on a system with at least four cores the tests WILL run in multiple threads and hence have to be synchronized properly. Will create a PR.

**Comment by ArneDeutsch** [comment]
After investigating more tests I became unsure ... the ClockUtil is used all over the test code base. To my understanding this implies, that most of the tests are inherintly not threadsafe. It is not possible (and probably not advisable) to synchronize them all ... hence I would think it is better to run the tests in one thread only.

Here is a random example. I just wonder if another test that uses the ClockUtil is executed concurrently ... I would think that it messes with this one.

```
  @Test
  public void occurredAfterParameterWorks() {
    // given
    Date now = new Date();
    ClockUtil.setCurrentTime(now);
    final ProcessInstance processInstance = engineRule.getRuntimeService().startProcessInstanceByKey("process");
    runtimeService.suspendProcessInstanceById(processInstance.getProcessInstanceId());

    Date nowPlus2Seconds = new Date(now.getTime() + 2000L);
    ClockUtil.setCurrentTime(nowPlus2Seconds);
    runtimeService.activateProcessInstanceById(processInstance.getProcessInstanceId());

    Date nowPlus4Seconds = new Date(now.getTime() + 4000L);
    ClockUtil.setCurrentTime(nowPlus4Seconds);
    runtimeService.suspendProcessInstanceById(processInstance.getProcessInstanceId());

    // when
    List<UserOperationLogEntry> userOperationsLog =
      optimizeService.getHistoricUserOperationLogs(now, null, 10);

    // then
    Set<String> allowedOperationsTypes = new HashSet<>(Arrays.asList(OPERATION_TYPE_SUSPEND, OPERATION_TYPE_ACTIVATE));
    assertThat(userOperationsLog).hasSize(2);
    assertTrue(allowedOperationsTypes.contains(userOperationsLog.get(0).getOperationType()));
    assertTrue(allowedOperationsTypes.contains(userOperationsLog.get(1).getOperationType()));
  }
```


Thoughts?

**Comment by ArneDeutsch** [comment]
Traced this back to https://github.com/operaton/operaton/issues/205 . Here the multithreading behaviour was introduced. For me the only solution is change back

`<surefire.forkCount>0.5C</surefire.forkCount>`

to

`<surefire.forkCount>1</surefire.forkCount>`

Of course this will degrade build performance but I see no other way to ensure all the test work without race conditions.

**Comment by javahippie** [comment]
Hi @ArneDeutsch, thanks for investigating this!

I tend to agree. The ClockUtil is not the only reason why the tests cannot be run in parallel, there are many cases where a centralized test setup is done, which is shared between all tests. I'd be fine with setting the forkCount back to 1 again

**Comment by kthoms** [comment]
I would be happy if this situation could be gradually improved. Tests should be executable in parallel, and the fact that we can't is significantly influencing the build time.
I would be very happy if you could identify and solve such issues.

**Comment by javahippie** [comment]
@kthoms I agree, that is a huge task which will require a lot of heavy lifting, though.

Should we close this issue? I just merged the PR by @ArneDeutsch and think that his assumptions are very plausible. I have not seen this issue now for some time, and it might be sensible to assume it closed.

**Comment by ArneDeutsch** [comment]
To make these tests thread safe one would (at least) need to convert _ClockUtil_ from a static helper utility to some instance hold somewhere in the processing engine. On some central place that is instantiated once per test case. Such that there is exactly one timing instance per test case. Everyone that needs the time would then need to access the same instance. Honestly I have no idea if this is feasable and where such an instance could be living.

**Comment by ArneDeutsch** [comment]
I think this ticket can be closed because the test should be good now. We could open a new ticket to make tests able to be executed in multiple threads.
